DCOS-42520: UI Update Service: Systemd setup

GeorgiSTodorov opened the pull request:
Add minimal systemd setup for the ui update service.

Closes https://jira.mesosphere.com/browse/DCOS-42520

## High-level description

What features does this change enable? What bugs does this change fix?

_It is a step in creating DCOS UI Update Service_

## Corresponding DC/OS tickets (obligatory)

These DC/OS JIRA ticket(s) must be updated (ideally closed) in the moment this PR lands:

  - https://jira.mesosphere.com/browse/DCOS-42520


## Related tickets (optional)

Other tickets related to this change:

  - https://jira.mesosphere.com/browse/DCOS-42493


## Checklist for all PRs

  - [x] Added a comprehensible changelog entry to `CHANGES.md` or explain why this is not a user-facing change: _Just a minimal setup, not a user-facing change_
  - [x] Included a test which will fail if code is reverted but test is not. If there is no test please explain here: _This is just a minimal setup, we can test it once it is complete_
  - [x] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [x] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)


## Checklist for component/package updates:

If you are changing components or packages in DC/OS (e.g. you are bumping the sha or ref of anything underneath `packages`), then in addition to the above please also include:

  - [ ] Change log from the last version integrated (this should be a link to commits for easy verification and review): [example](https://github.com/dcos/dcos-mesos-modules/compare/f6fa27d7c40f4207ba3bb2274e2cfe79b62a395a...6660b90fbbf69a15ef46d0184e36755881d6a5ae)
  - [ ] Test Results: [link to CI job test results for component]
  - [ ] Code Coverage (if available): [link to code coverage report]
___

GeorgiSTodorov commented:
I don't have permissions to see why some tests fail.

gpaul commented:
That's a good start. I think you should compare your package with e.g., mesos-dns: 
https://github.com/dcos/dcos/tree/master/packages/mesos-dns

You need `buildinfo.json` and `build` files - you should probably _not_ specify a custom build docker - those tend to get left behind as we update packages and versions in our main build docker.

gpaul commented:
Instead, consider updating the Go version used throughout the project to 1.11.1 here:
https://github.com/dcos/dcos/blob/master/pkgpanda/docker/dcos-builder/Dockerfile#L53

Let's discuss updating the DC/OS Go version in the #go slack channel if you have access to the mesosphere internal slack.

GeorgiSTodorov commented:
@gpaul Thank you a lot for your suggestions.

I have added the `build` and `buildinfo.json` files. I have also changed the Golang version. A few questions:
- Is everything ok with my `build` and `buildinfo.json` files?
- Do I need to change the `ref` in `buildinfo.json` and Golang SHA256 and if yes, how?

gpaul commented on packages/dcos-ui-update-service/extra/dcos-ui-update-service.service:
```diff
@@ -0,0 +1,14 @@
+EnvironmentFile=-/run/dcos/etc/dcos-ui-update-service.env
+ExecStartPre=/opt/mesosphere/bin/bootstrap dcos-ui-update-service
+ExecStart=/opt/mesosphere/bin/dcos-ui-update-service --verbose=true
+Sockets=dcos-ui-update-service.socket
```
I'm not sure I would use systemd to manage your socket file unless you really need that. For starters I would remove the `Sockets=` directive and the `*.socket` file and just tell your program to listen on a given network (`unix`) and address (`/run/dcos/dcos-ui-update-service.sock`).

gpaul replied:
Hmm - maybe sockets are actually better - they let you specify the file ownership, which is nice for restricting access to your service from Admin Router. In that case, you'll want to use https://github.com/coreos/go-systemd/tree/master/activation to obtain the listener that your server listens on.

GeorgiSTodorov replied:
How to obtain it?

gpaul replied:
Here's a better example than mesos-dns:
https://github.com/mesosphere/dcos-enterprise/tree/master/packages/dcos-backup/extra

It's pretty similar to what you have right now.

For a code sample of how this is used see:
https://github.com/mesosphere/dcos-backup/blob/master/main.go#L45-L63

GeorgiSTodorov replied:
I have no access to the second repository, I will check the first though.

gpaul replied:
Let's get you that access.

gpaul replied:
Right - the socket code you have is perfect - you now need to adjust your Go program to be able to receive connections through systemd. That's what the link to dcos-backup is for, it serves as a solid example of how to do that. I asked to have you granted access to that repository, I'm not sure whether you have it yet. Please follow up on that.

GeorgiSTodorov replied:
Please check https://github.com/dcos/dcos/pull/3566/commits/dda368074f37499991d12c0884177419751438e8 to see if I did it right. I still don't have access to that repository, I guess it might take some time.

gpaul replied:
I'll ping J now.

GeorgiSTodorov replied:
I have access now.

gpaul commented on pkgpanda/docker/dcos-builder/Dockerfile:
```diff
@@ -50,7 +50,7 @@ RUN ln -sf /usr/bin/cpp-4.8 /usr/bin/cpp && \
-ENV GOLANG_VERSION 1.8.3
+ENV GOLANG_VERSION 1.11.1
 ENV GOLANG_DOWNLOAD_URL https://golang.org/dl/go$GOLANG_VERSION.linux-amd64.tar.gz
 ENV GOLANG_DOWNLOAD_SHA256 1862f4c3d3907e59b04a757cfda0ea7aa9ef39274af99a784f5be843c80c6772
```
You'll need to update the sha256sum to match the downloaded file. It is available on the golang.org/dl page.

nLight commented:
superseded by https://github.com/dcos/dcos/pull/3618
@orsenthil could you please close it?

nLight commented:
Hey @orsenthil would you mind closing the PR? Thanks!
